implemented dao methods

oscarwang20 opened the pull request:
### Summary <!-- Required -->

<!-- Provide a general summary of your changes in the Title above -->
<!-- Itemize bug fixes, new features, and other changes -->
<!-- Feel free to break this into sub-sections, i.e. features, fixes, etc. -->
<!-- Some examples are shown below. -->

Implemented dao methods to:
- delete all team event attendance
- get all team event attendance by user
- get all team event attendance by event
- get all team event attendance

### Notion/Figma Link <!-- Optional -->

https://www.notion.so/cornelldti/Implement-deleteAllTeamEventAttendance-and-get-dao-methods-for-TeamEventAttendanceDao-6e9407c1e1834c28af665951c975a3ca?pvs=4

### Test Plan <!-- Required -->

<!-- Provide screenshots or point out the additional unit tests -->
Tests are from the following file:
```
import { fakeIdolMember } from '../../tests/data/createData';
import { teamEventAttendanceCollection } from '../firebase';
import TeamEventAttendanceDao from './TeamEventAttendanceDao';

const member1 = {
  netid: 'ow39',
  email: 'ow39@cornell.edu',
  firstName: 'Oscar',
  lastName: 'Wang',
  pronouns: '',
  graduation: '',
  major: '',
  hometown: '',
  about: '',
  subteams: '',
  role: 'developer',
  roleDescription: 'Developer'
};
const member2 = fakeIdolMember();

const testAttendance1: TeamEventAttendance = {
  member: member1,
  hoursAttended: 1,
  image: '',
  eventUuid: 'event1',
  pending: true,
  uuid: 'test1'
};

const testAttendance2: TeamEventAttendance = {
  member: member2,
  hoursAttended: 1,
  image: '',
  eventUuid: 'event1',
  pending: true,
  uuid: 'test2'
};

const testAttendance3: TeamEventAttendance = {
  member: member1,
  hoursAttended: 1,
  image: '',
  eventUuid: 'event2',
  pending: true,
  uuid: 'test3'
};

TeamEventAttendanceDao.createTeamEventAttendance(testAttendance1).then((res) => console.log(res));
TeamEventAttendanceDao.createTeamEventAttendance(testAttendance2).then((res) => console.log(res));
TeamEventAttendanceDao.createTeamEventAttendance(testAttendance3).then((res) => console.log(res));
TeamEventAttendanceDao.getAllTeamEventAttendance().then((res) => console.log(res));
TeamEventAttendanceDao.getTeamEventAttendance('event1').then((res) => console.log(res));
TeamEventAttendanceDao.getAllUsersTeamEventAttendance(member1).then((res) => console.log(res));
TeamEventAttendanceDao.deleteAllTeamEventAttendance().then((res) => console.log(res));
```

Get all team event attendance
<img width="644" alt="image" src="https://user-images.githubusercontent.com/69322572/226491255-a91783b4-c21d-474d-b250-6f5b3fde3018.png">

Get team event attendance by event
<img width="645" alt="image" src="https://user-images.githubusercontent.com/69322572/226491510-5d649b84-0fcd-48a0-b469-50ba8d0dd9d3.png">

Get team event attendance by user
<img width="627" alt="image" src="https://user-images.githubusercontent.com/69322572/226491611-3c92f04e-33f4-4a4e-8c44-d6de75d07fb4.png">

henry-li-06 commented on backend/src/dao/TeamEventAttendanceDao.ts:
```diff
@@ -51,4 +52,67 @@ export default class TeamEventAttendanceDao {
+    const attendanceRefs = await teamEventAttendanceCollection.get();
+    return Promise.all(
+      attendanceRefs.docs.map(async (doc) => {
+        const { eventUuid, hoursAttended, image, member, pending, uuid } = doc.data();
+        return {
+          eventUuid,
+          hoursAttended,
+          image,
+          member: await getMemberFromDocumentReference(member),
+          pending,
+          uuid
+        };
```
You can just use the spread operator on `doc.data()` for most of the values.

henry-li-06 commented on backend/src/dao/TeamEventAttendanceDao.ts:
```diff
@@ -51,4 +52,67 @@ export default class TeamEventAttendanceDao {
+   * @param teamEventAttendance - updated TeamEventAttendance object
+   *  If hoursAttended is originally defined, it cannot be overwritten afterwards
+   */
+  static async getAllUsersTeamEventAttendance(user: IdolMember): Promise<TeamEventAttendance[]> {
```
I think it might be better to name this function `getTeamEventAttendanceByUser`

henry-li-06 commented on backend/src/dao/TeamEventAttendanceDao.ts:
```diff
@@ -51,4 +52,67 @@ export default class TeamEventAttendanceDao {
+   * Deletes a TEC Attendance
+   * @param uuid - DB uuid of TeamEventAttendance
+   */
+  static async getTeamEventAttendance(uuid: string): Promise<TeamEventAttendance[]> {
```
Might be better to name this function `getTeamEventAttendanceByEventID`

henry-li-06 reviewed (commented):
Some of the method specs don't seem to match the actual method.

oscarwang20 commented:
> Some of the method specs don't seem to match the actual method.

Just pushed the changes!

henry-li-06 reviewed (approved):
LGTM
```
                 |~
           |/    w
          / (   (|   \
         /( (/   |)  |\
  ____  ( (/    (|   | )  ,
 |----\ (/ |    /|   |'\ /^;
\---*---Y--+-----+---+--/(
 \------*---*--*---*--/
  '~~ ~~~~~~~~~~~~~~~
```

andrew032011 commented on backend/src/dao/TeamEventAttendanceDao.ts:
```diff
@@ -51,4 +52,59 @@ export default class TeamEventAttendanceDao {
+    const coll = await teamEventAttendanceCollection.get();
+
+    coll.docs.forEach((doc) => batch.delete(doc.ref));
+    await batch.commit();
```
This logic looks mostly correct. Just want to double check with @henry-li-06 @JacksonStaniec though. Does the batch get all the documents within the collection, or just a selection of them? I wonder if we tried this on a larger amount of data (i.e. 500+ documents), if this would fail. 

Reading up [here](https://cloud.google.com/firestore/quotas#writes_and_transactions). 

> 

Maximum number of writes that can be passed to a Commit operation or performed in a transaction | 500
-- | --




If the batch ends up going over 500, this delete would fail, since we can't make more than 500 writes at once. Also, the code assumes that the doc is within the batch. What happens when we call `batch.delete(...)` and the document is not in the batch? We don't have a fallback for that.

This wouldn't really be a problem for a bit, especially since we're not going to use this method for awhile, but I can see us getting up to 500 fast. If you don't end up making any changes here, I would at the very least mention in the specification that this could fail if there's 500+ documents, and we'd file a ticket to come back and fix this.

I found some recipes here for deleting all documents in a collection, one of them takes into account the 500 limit and splits up the batches into max 500 documents each: 

https://firebase.google.com/docs/firestore/manage-data/delete-data#collections

https://stackoverflow.com/questions/48175235/how-to-delete-multiple-documents-from-cloud-firestore

andrew032011 replied:
Ah, great @oscarwang20 ignore the solutions I posted then! Looks like we already have a solution (which happens to be taken directly off the Firebase solution I linked).

andrew032011 commented on backend/src/dao/TeamEventAttendanceDao.ts:
```diff
@@ -51,4 +52,59 @@ export default class TeamEventAttendanceDao {
+    return Promise.all(
+      attendanceRefs.docs.map(async (doc) => ({
+        ...doc.data(),
+        member: await getMemberFromDocumentReference(doc.get('member'))
```
I don't think we need to call `get` from the doc every time. `doc.data()` should have the member information. I believe your logic here would double the amount of reads for each member (which we've been working to reduce these days). Since it's not too complicated to optimize here I would add that: save doc.data() in a variable (of type `DBTeamEventAttendance`) and then grab the `member` field off of that.

andrew032011 reviewed (changes requested):
Nice work! See my comments.

I see you also changed the names of some of the dao methods. That's fine with me. I'm curious, did you work with Jackson on renaming those? If we agree on some name changes, it would be good to update it in the ticket as well, so it's well documented. And I can make updates in the design doc too (since it might affect other tickets).

henry-li-06 commented:
> Nice work! See my comments.
> 
> I see you also changed the names of some of the dao methods. That's fine with me. I'm curious, did you work with Jackson on renaming those? If we agree on some name changes, it would be good to update it in the ticket as well, so it's well documented. And I can make updates in the design doc too (since it might affect other tickets).

I asked for the method names to be changed.

andrew032011 commented:
> > Nice work! See my comments.
> > I see you also changed the names of some of the dao methods. That's fine with me. I'm curious, did you work with Jackson on renaming those? If we agree on some name changes, it would be good to update it in the ticket as well, so it's well documented. And I can make updates in the design doc too (since it might affect other tickets).
> 
> I asked for the method names to be changed, the original names were very ambiguous.

Oh right, I see your comments above. I agree with those changes then. Once we merge this in I can update those names in the ticket + design doc + other tickets.

andrew032011 commented:
> > > > Nice work! See my comments.
> > > > I see you also changed the names of some of the dao methods. That's fine with me. I'm curious, did you work with Jackson on renaming those? If we agree on some name changes, it would be good to update it in the ticket as well, so it's well documented. And I can make updates in the design doc too (since it might affect other tickets).
> > > 
> > > 
> > > I asked for the method names to be changed, the original names were very ambiguous.
> > 
> > 
> > Oh right, I see your comments above. I agree with those changes then. Once we merge this in I can update those names in the ticket + design doc + other tickets.
> 
> @andrew032011 I also like the new names. Will we update the names of the corresponding API endpoints as well?

I’ll revisit those tonight! If there’s updates I will let you know!

andrew032011 commented:
Could you add the solution @henry-li-06 posted to delete the collection? I meant ignore the solutions I posted and use this code instead: https://github.com/cornell-dti/idol/blob/main/backend/scripts/update-dev-db.ts#L17-L56, since we already have a solution implemented for delete a collection correctly.

andrew032011 commented:
> > > > Nice work! See my comments.
> > > > I see you also changed the names of some of the dao methods. That's fine with me. I'm curious, did you work with Jackson on renaming those? If we agree on some name changes, it would be good to update it in the ticket as well, so it's well documented. And I can make updates in the design doc too (since it might affect other tickets).
> > > 
> > > 
> > > I asked for the method names to be changed, the original names were very ambiguous.
> > 
> > 
> > Oh right, I see your comments above. I agree with those changes then. Once we merge this in I can update those names in the ticket + design doc + other tickets.
> 
> @andrew032011 I also like the new names. Will we update the names of the corresponding API endpoints as well?

Endpoint `/getAllUsersTeamEventAttendance` has been renamed to `/getTeamEventAttendanceByUser`. Updated in design doc and your ticket.

Other dao method name updates have also propagated through the tickets and design doc, so everything should be up to date now.

oscarwang20 commented:
- Implemented `deleteAllTeamEventAttendanceHelper` to delete 500 documents at a time
- refactored `deleteAllTeamEventAttendance` to recursively call `deleteAllTeamEventAttendanceHelper` when there are still documents to delete after each batch deletion

andrew032011 commented:
Feel free to message me on Slack if it’s not clear what I mean.

oscarwang20 commented:
- moved `deleteCollection` method to `firebase-utils.ts`
- `update-dev-db.ts` and `teamEventAttendanceDao.ts` now both import `deleteCollection` from `firebase-utils.ts`

andrew032011 commented on backend/src/utils/firebase-utils.ts:
```diff
@@ -16,4 +17,45 @@ const configureAccount = (sa: any, isProd: boolean) => {
+  });
+}
+
 export default configureAccount;
```
I'd remove this line and update the imports that are affected.

henry-li-06 commented on backend/src/dao/TeamEventAttendanceDao.ts:
```diff
@@ -58,26 +59,12 @@ export default class TeamEventAttendanceDao {
+    await deleteCollection(db, 'team-event-attendance', 500);
     if (snapshot.size !== 0) {
       await this.deleteAllTeamEventAttendance();
     }
```
I don't think this is needed anymore

andrew032011 replied:
+1. The logic in `deleteCollection` already takes care of ensuring that all the documents are deleted. You specify the batch size for the number in each batch that we delete (500 is the limit), and the function will keep deleting in batches of size 500. You don't need to recurse further.
